docs/batch_size_optimizer/server.md

```diff
@@ -0,0 +1,156 @@
+# Batch Size Optimizer in Zeus
+
+## What is it?
+
+Batch size optimizer(BSO) can choose the best batch size that minimizes the cost, where cost is defined as $cost = \eta \times \text{Energy consumption to accuracy} + (1-\eta) \times \text{Max power}\times \text{Time to accuracy}$.
+
+## How does it work?
+
+Core of BSO is a multi-arm-bandit based on **recurrent** trainings. After each training, we feed the result cost to MAB and after certain number of trainings, MAB can converge to the best batch size. In addition to MAB, we employed early-stopping and pruning to handle stragglers. For more detail, refer to [paper](https://www.usenix.org/conference/nsdi23/presentation/you).
+
+## Should I use this?
+
+The key of BSO is recurrent training. If you are training your model periodically or repeatedly, BSO can be a great choice to reduce energy or time consumption.
+
+## Limitations
+
+We currently doesn't support heterogeneous GPUs or different configurations. Number of GPUs, gpu models, and other configurations in JobSpec should be identical in recurrent trainings. If you are running your training in a various environment each time, then it might not desirable to use BSO.
+
+## Quick start (Server)
+
+1. Clone the repository
+
+    ```Shell
+    git clone https://github.com/ml-energy/zeus/tree/master
+    ```
+
+2. Create `.env` under `/docker`. Example of `.env` is provided below.
+
+    For default, we are using the MySQL for the database.
+
+    ```Shell
+    ZEUS_BSO_DB_USER="me" 
+    ZEUS_BSO_DB_PASSWORD="secret"
+    ZEUS_BSO_ROOT_PASSWORD="secret*"
+    ZEUS_BSO_SERVER_PORT=8000
+    ZEUS_BSO_LOG_LEVEL = "INFO"
+    ZEUS_BSO_ECHO_SQL = "True"
+    ```
+
+    If you want to use different databases, you need to add `ZEUS_BSO_DATABASE_URL` as an environment variable. See [Remark](#remark-about-server) for detail.
+    Also, if you are running using docker-compose or kubernetes, you need to change the image name under `db` in docker-compose file.
+
+3. Running a server
+    - Using docker-compose
+
+        ```Shell
+        cd docker 
+        docker-compose -f ./docker/docker-compose.yaml up
+        ```
+
+        This will build images for each container: db, migration and the server. Then, it will spin those containers.
+
+    - Using Kubernetes.
+        1. Build an image.
+
+            ```Shell
+            # From the root directory
+            build -f ./docker/server.Dockerfile -t bso-server . 
+            build -f ./docker/migration.Dockerfile -t bso-server .
```

Comment:
```suggestion
            docker build -f ./docker/server.Dockerfile -t bso-server . 
            docker build -f ./docker/migration.Dockerfile -t bso-server .
```

1. Why is there a separate Docker container for migration? Is it basically one-time use just for migration? If so, it should be mentioned. Also, the command to actually run migration should be there.
2. Tag names are the same. Mistake?

Reply:
1. It is a one-time thing as long as our schema doesn't change. But it records the version of schema at each time it runs so I think it is good to run at each deployment. However, for our case doesn't really matter. In terms of the reason why using a container, I read a couple of articles saying running migrations in a container is p good practice since it simplifies deployment. 
2. Yep fixed

Reply:
Does the migration container also generates files that need to be used by the server container? Or is it just directly talking to the DB container?

Reply:
The migration container generates the version script each time it performs migration. In terms of applying the migration, it directly talks to the DB container.